pymc/util.py

```diff
@@ -231,38 +230,26 @@ def enhanced(*args, **kwargs):
     return enhanced
 
 
-def dataset_to_point_list(ds: xarray.Dataset) -> List[Dict[str, np.ndarray]]:
+def dataset_to_point_list(
+    ds: xarray.Dataset, sample_dims: List
+) -> Tuple[List[Dict[str, np.ndarray]], Dict[str, Any]]:
     # All keys of the dataset must be a str
-    for vn in ds.keys():
+    var_names = list(ds.keys())
+    for vn in var_names:
         if not isinstance(vn, str):
             raise ValueError(f"Variable names must be str, but dataset key {vn} is a {type(vn)}.")
-    # make dicts
-    points: List[Dict[Hashable, np.ndarray]] = []
-    da: "xarray.DataArray"
-    for c in ds.chain:
-        for d in ds.draw:
-            points.append({vn: da.sel(chain=c, draw=d).values for vn, da in ds.items()})
+    num_sample_dims = len(sample_dims)
+    stacked_dims = {dim_name: ds[dim_name] for dim_name in sample_dims}
+    ds = ds.transpose(*sample_dims, ...)
```

Comment:
I’ma bit confused here. Where does the speed up come from? Does it come from not appending to a list?
If yes, then transposing and creating the intermediate dictionaries seems unnecessary to me.

Could you do something like
```
points = [
    {vn: da.values for vn, da in ds.sel(**coords).items()}
    for coords in [
        dict(stacked_dims, coord) for coord in
        itertools.product(stacked_dims.values())
    ]
]
```

I still feel that this could be speed up more if we could avoid the need to create dictionaries and use datasets as mappings. Anyway, if I misunderstood where the speed up is coming from, feel free to ignore this

Reply:
IIRC the speedup comes from doing numerical indexing in the inner loop as opposed to dim indexing.

Reply:
If that’s the case, we could use `isel` instead of `sel` in my example to also use numerical indexing

Reply:
Maybe, if that's as fast as nunpy indexing. Internet tells me xarray operations of this kind have a lot of overhead, maybe isel doesn't...

Reply:
It looks like xarray's overhead is huge regardless

![image](https://user-images.githubusercontent.com/5230109/195785834-1df2b671-0dde-446c-bb37-cf06127b57bb.png)

`isel` is ~500x slower than direct indexing of the numpy array
`sel` is ~1000x slower than direct indexing of the numpy array

So, with 4000 chains and draws, xarray would take seconds to finish with this small test `DataArray`, whereas indexing directly with numpy would cut that down to milliseconds.

Reply:
I was curious and actually tested what I said above

![image](https://user-images.githubusercontent.com/5230109/195788201-209cf71c-1489-4c25-bbd3-287937f20ac8.png)

In the end, xarray was faster than I had anticipated, but it's still more than 100x slower than numpy. I'm just amazed at how slow iterating with xarray is actually! I thought that they would have at least tried to stay close to numpy.

Reply:
> sel is ~1000x slower than direct indexing of the numpy array

Yeah Oriol got the same numbers (1000x) when he timed it before opening the PR

Reply:
Yeah, I think xarray here is more like pandas than numpy, it is explicitly not designed for this kind of thing but to work with objects as a whole or take "array subsets". Both `i/sel(chain=0)` probably check `chain` exists, then find which position it is in the underlying array, generate the whole `0, :` or `:, 0, :, :` to be used with numpy and then call numpy.

As for reshaping, I am under the impression that subsetting over a single axis is marginally faster than over 2 or 3 axis so given how we potentially repeat that same subset a few thousand times tipically it might help a bit once we are in ms range.